src/board/square.rs

```diff
@@ -0,0 +1,59 @@
+use super::{Board, errors::InvalidSquareError, Orientation};
+
+// Represents a square on the board
+//
+// File counts from the left, starts at 0
+// Rank counts from the bottom, starts at 0
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct Square {
+    pub file: usize,
+    pub rank: usize,
+}
+
+impl Square {
+    pub fn get_file(&self) -> usize {
+        return self.file;
+    }
+
+    pub fn get_rank(&self) -> usize {
+        return self.rank;
+    }
+}
+
+// Represents a possibly invalid location of a square
+// on the board.
+// File counts from the left, starts at 0
+// Rank counts from the bottom, starts at 0
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct UncheckedSquare {
+    pub file: usize,
+    pub rank: usize,
+}
+
+impl UncheckedSquare {
+    // Checks that the square is valid on the given board
+    // TODO: validate through the type system that it's valid for only the board
+    // that it's been tested on. This may require some lifetime shenanigans
+    pub fn validate(self, board: &Board) -> Result<Square, InvalidSquareError> {
```

Comment:
I'm wondering if this shouldn't be a part of `Board` rather than here. `unchecked_square.validate(board)?` reads worse than `board.check(unchecked_square)?` or something. I also think that `check` is a good function name here, since it's already in the name: it's `Unchecked`, so you clearly `check` it to fix it.